### ISSUE TITLE
allow outer PEP 695 typevars rescoped to Callable return to be usable inside function body

### ISSUE THREAD

**Issue by gjcarneiro** [astral-sh/ty#2723]
### Question

This code used to work in ty 0.0.14, but now I get errors.  Any hint on how to correctly annotate this for Ty would be appreciated.

I mean the types P and R to be the same type throughout all the layers of nesting, but Ty insists that they are different types.  I am baffled.

```py
from typing import Callable
from functools import wraps


def monitored_external_call[**P, R](
    api_name: str,
    api_call: str | None = None,
) -> Callable[[Callable[P, R]], Callable[P, R]]:

    def outer(func: Callable[P, R]) -> Callable[P, R]:
        @wraps(func)
        def inner(*args: P.args, **kwargs: P.kwargs) -> R:
            return func(*args, **kwargs)

        return inner

    return outer
```

Output:

```
    Return type does not match returned value: expected `R@inner`, found `R@outer` (invalid-return-type) [Ln 13, Col 20]
    Argument is incorrect: Expected `P@outer.args`, found `P@inner.args` (invalid-argument-type) [Ln 13, Col 25]
    Argument is incorrect: Expected `P@outer.kwargs`, found `P@inner.kwargs` (invalid-argument-type) [Ln 13, Col 32]
    Return type does not match returned value: expected `(**P@outer) -> R@outer`, found `_Wrapped[P@outer, R@outer, P@inner, R@inner]` (invalid-return-type) [Ln 15, Col 16]
    Return type does not match returned value: expected `[**P'return, R'return]((**P'return@monitored_external_call) -> R'return, /) -> (**P'return@monitored_external_call) -> R'return`, found `def outer(func: (**P@outer) -> R@outer) -> (**P@outer) -> R@outer` (invalid-return-type) [Ln 17, Col 12]
```

### Version

ty 0.0.15

**Comment by sharkdp** [comment]
cc @dcreager

**Comment by gjcarneiro** [comment]
Huh, I think I solved it, sort of:

```diff
-    def outer(func: Callable[P, R]) -> Callable[P, R]:
+    def outer[**P, R](func: Callable[P, R]) -> Callable[P, R]:
```

That makes type checking work.  To me, it seems illogical, but Ty seems to accept it, and seems to do the right thing... 🤷 

I'll close it for now, but, if you guys think this is a bug, feel free to reopen.

**Comment by dcreager** [comment]
This is very subtle! And the tl;dr is that pyright [supports this](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgLABQokUwArigMYxhgkDOmbLnwB3EEQR8G0%2BgBNqwKBDAok3ENVkB9agA8Y1EClJaOpEgG0AVFYAKAGigAlALoAKBlC9QJSLcYhqAC4oPhgQe09vX1NzELCQKAAfKAA5FWooAF40jMj6AEooAFoAPkJzShoLC2IyKuoLB2cXF0c68ipG5tcXIJlvKHlFMFYDEDc2ThCOhqbHVyKyivqu%2BZb%2B%2BkHBgAExCT5J9g4CqO2hhUwUFEM3KyIQND4Q2wA6B6fHGwBrEQ-nqBvX7-JblJybc7nDQwVhGFjHO7-L5WYGPPinehnbbQ2EoK43EADbw4uGjcYyYaYeQoGBqOBuXQhVAwUFXGAQ4nUGFw3QyDQAN2oJngCGobmUqnUmh0%2BkMxhIsTIbgARMAeMrHMqKA9lQUMQKhQqRWKJWpcNK9ON5YqSCq1WANVAtTqCm4kNTafA9QwgA) as you originally wrote it, so we should probably consider supporting that too.

To explain the behavior you were seeing: (apologies if this is too much detail, just wanting to write it all down for the record)

You have a "decorator factory" (`monitored_external_call`), which creates and returns a new decorator (`outer`) for each `api_name` and `api_call`. `outer` is a decorator, which takes in a function, decorates it, and returns the decorated function (`inner`).

The decorator needs to be generic, since it should be able to work on any function. That's why it takes in (and returns) a `Callable[P, R]`. The decorated function (`inner`) is _not_ generic — by the time we're creating and returning an `inner`, we have a specific function that we're decorating. So even though we see `P.args`, `P.kwargs`, and `R` in `inner`'s signature, it is _not_ generic. There is a separate `inner` created for each particular `P` and `R`, each time we call `outer`.

The decorator factory is also not generic, since it never sees the particular functions that we're decorating. It returns, _and then_ we use the returned decorator on however many functions we want to decorate.

So to recap:

- `monitored_external_call` should not be generic, but should return a generic `Callable`
- `outer` should be
- `inner` should not be

Unfortunately, Python does not have a convenient syntax to say that `monitored_external_call` is not generic, but returns a generic callable.

We merged https://github.com/astral-sh/ruff/pull/22954 in the last couple of days which adds a heuristic for this. We notice that `monitored_external_call` binds two typevars (`P` and `R`), but that those typevars are only mentioned in the `Callable` that it returns. When we see this pattern, we interpret it as described above: the function is not generic, and the `Callable` that it returns is.

Okay, that's all well and good. One last wrinkle is that, with the way you originally wrote it, it looks like `outer` is not generic! It refers to the `P` and `R` from `monitored_external_call`. If `monitored_external_call` were generic, then this pattern makes `outer` _not_ generic — we'd create a new `outer` for the particular instantiation of `P` and `R` each time `monitored_external_call` is called. (Just like happens with `outer`/`inner` in the "correct"/intended behavior.) Adding the `[**P, R]` to `outer` is the correct way to state that `outer` is generic, regardless of whether `monitored_external_call` is.

So, I would argue that what you changed it to is a more accurate description of which of the functions should be generic and which should not be. I agree that it's confusing though! Especially since pyright accepts your original formulation.

At the very least, we can update our diagnostics to recognize this situation and suggest the edit that you made. We can also consider matching pyright's interpretation, which still considers `P` and `R` to be in scope inside of the body of `monitored_external_call`, even if we've decided it's not generic. (That would make us treat those PEP-695 typevars the same way that we would treat legacy typevars.)

**Comment by AlexWaygood** [comment]
@dcreager should we reopen this to track those improvements?

**Comment by carljm** [comment]
Pyright, mypy, and pyrefly all accept the original formulation, so yeah I think we should reopen this and do that as well.

**Comment by JulianJvn** [comment]
In addition, mypy flags the proposed addition of `[**P, R]` to `outer()` with `"P [or R]" already defined as a type parameter  [misc]`. Might be a weak argument to say ty should allow the code without `[**P, R]`, but it does mean that one has to add additional ignore comments to repos that use both mypy and ty as long as ty is in beta.

**Comment by sharifhsn** [comment]
Closed by https://github.com/astral-sh/ruff/pull/24909

Repro:

```py
from typing import Callable
from functools import wraps

def monitored_external_call[**P, R](
    api_name: str,
    api_call: str | None = None,
) -> Callable[[Callable[P, R]], Callable[P, R]]:
    def outer(func: Callable[P, R]) -> Callable[P, R]:
        @wraps(func)
        def inner(*args: P.args, **kwargs: P.kwargs) -> R:
            return func(*args, **kwargs)
        return inner
    return outer
```

ty 0.0.35:

```
error[invalid-return-type]: Return type does not match returned value
  --> 2723.py:11:57
   |
11 |         def inner(*args: P.args, **kwargs: P.kwargs) -> R:
   |                                                         - Expected `R@inner` because of return type
12 |             return func(*args, **kwargs)
   |                    ^^^^^^^^^^^^^^^^^^^^^ expected `R@inner`, found `R@outer`

error[invalid-argument-type]: Argument is incorrect
  --> 2723.py:12:25
   |
12 |             return func(*args, **kwargs)
   |                         ^^^^^ Expected `P@outer.args`, found `P@inner.args`

Found 5 diagnostics
```

ty 0.0.36:

```
All checks passed!
```
